driver/src/solution_submission/mod.rs

```diff
@@ -120,48 +128,189 @@ impl<'a> StableXSolutionSubmitting for StableXSolutionSubmitter<'a> {
         solution: Solution,
         claimed_objective_value: U256,
     ) -> Result<(), SolutionSubmissionError> {
-        self.contract
-            .submit_solution(
-                batch_index.clone(),
-                solution.clone(),
-                claimed_objective_value,
+        retry_with_gas_price_increase(
+            self.contract,
+            batch_index,
+            solution.clone(),
+            claimed_objective_value,
+            self.gas_price_estimating,
+            60_000_000_000u64.into(),
+        )
+        .map_err(|err| {
+            extract_transaction_receipt(&err)
+                .and_then(|tx| {
+                    let block_number = tx.block_number?;
+                    match self.contract.get_solution_objective_value(
+                        batch_index,
+                        solution,
+                        Some(block_number.into()),
+                    ) {
+                        Ok(_) => None,
+                        Err(e) => Some(SolutionSubmissionError::from(e)),
+                    }
+                })
+                .unwrap_or_else(|| SolutionSubmissionError::Unexpected(err.into()))
+        })
+        .map(|_| ())
+    }
+}
+
+fn retry_with_gas_price_increase(
+    contract: &dyn StableXContract,
+    batch_index: U256,
+    solution: Solution,
+    claimed_objective_value: U256,
+    gas_price_estimating: &dyn GasPriceEstimating,
+    gas_cap: U256,
+) -> Result<(), MethodError> {
+    const INCREASE_FACTOR: u32 = 2;
+    const BLOCK_TIMEOUT: usize = 1;
+    const DEFAULT_GAS_PRICE: u64 = 15_000_000_000;
+
+    let mut gas_price = match gas_price_estimating.estimate_gas_price() {
+        Ok(gas_estimate) => gas_estimate.fast,
+        Err(ref err) => {
+            log::warn!(
+                "failed to get gas price from gnosis safe gas station: {}",
+                err
+            );
+            U256::from(DEFAULT_GAS_PRICE)
+        }
+    };
+    // Never exceed the gas cap.
+    gas_price = std::cmp::min(gas_price, gas_cap);
+
+    let mut result;
+    // the following block emulates a do-while loop
+    while {
+        // Submit solution at estimated gas price (not exceeding the cap)
+        result = contract.submit_solution(
+            batch_index,
+            solution.clone(),
+            claimed_objective_value,
+            gas_price,
+            if gas_price == gas_cap {
+                None
+            } else {
+                Some(BLOCK_TIMEOUT)
+            },
+        );
+        // Breaking condition for our loop
+        gas_price < gas_cap
+            && matches!(
+                result,
+                Err(MethodError {
+                    inner: ExecutionError::ConfirmTimeout,
```

Comment:
It might be useful to log this, at least in the beginning. I'm afraid we might unexpectedly always get a timeout when using `block_timeout == 1` or something unexpected like that.